### DISCUSSION TITLE
Kdump (kernel crash dump) redesign

### DISCUSSION THREAD

**Discussion by garrett** [cockpit-project/cockpit#19203] [Show and tell]
I'm currently working on a Kdump redesign. And thought having a discussion about it would be useful.

Right now, we have discussion happening on a a pull request, but I'd like for the discussion to continue after the PR is merge. Hence, opening it up here.

---

The discussion:

@liutgnu from the kdump team started it @ https://github.com/cockpit-project/cockpit/pull/19107#issuecomment-1680152357:

> Hi skobyda,
> 
> I'm Tao Liu from kdump team, came from [bz2215024](https://bugzilla.redhat.com/show_bug.cgi?id=2215024). Thanks a lot for your efforts on this. Since cockpit support is one subtask of bz2215024, and I have worked on the TUI part for sometime, I can share some of my thoughts:
> 
> 1) It will be better to have an automatic test check mechanism, as @marusak pointed out, after machine reboot and user re-enter the kdump page, a sign indicating if the test is successful or not will be good. Relying on user himself to check if the "test vmcore" been generated is inconvenient, there might be plenty of vmcores located in a nfs/ssh server for "nfs/ssh kdump", the manual checking will be a boring work.
> 
> 2) It may not be the final desgin, I planed to introduce 2 new shell command line options for kdumpctl: "kdumpctl test" and "kdumpctl test-check". "kdumpctl test" will generate a timestamp, which will be used to mark the vmcore that it belongs to a specifc test. "kdumpctl test-check" will check the kdump folder after reboot back, no matter if it's a local folder or a nfs/ssh remote folder, to see a vmcore with the specific timestamp string exist or not, which then be the indicator of the test is successful or not. Maybe you can call "kdumpctl test-check" in cockpit if it is easier.
> 
> 3) If the user modifies the configuration file of kdump, the prior test result will be invalid and notify he should retest based on the current kdump config. I hadn't figure out the tech details, maybe some file monitor service to be involved.
> 
> Anyway, it's currently an early design phase for me, we can share thoughts or discuss kdump APIs if you need some underlying support.

And then @skobyda (basically) agreed with all the points @ https://github.com/cockpit-project/cockpit/pull/19107#issuecomment-1680198871

Then @liutgnu added:

> OK, I will try to design the kdumpctl APIs in consideration of cockpit, hope it can be clean and easy to use. Just FYI, there is the repo which we used for implementing TUI "kdumpctl test/test-check" etc, currently just a POC in the ltao-dev branch.

And then I chimed in with more mockups @ https://github.com/cockpit-project/cockpit/pull/19107#issuecomment-1680815345 (some other mockups are above in that PR, for the test crash confirmation modal).

**Comment by garrett** [comment]
Current WIP mockup:

![](https://user-images.githubusercontent.com/10246/261054336-ab0f03e8-9af8-48ed-9acb-1a947b0971c7.png)

> It's still lacking a lot, such as: the modals for each option, the alert for when the service is out of sync (that is: if it's enabled and the service isn't running), other error cases, etc. — but the bulk of the direction to modernize the Kdump page (and bring it up to established patterns in both Cockpit and PatternFly) is there.

> Source: [Kdump.penpot in Kdump.zip](https://github.com/cockpit-project/cockpit/files/12360881/Kdump.zip) (to be imported in Penpot)

I'll work on refining it and adding more soon.

**Comment by skobyda** [comment]
Putting here newer mockup with me and @garrett discussed privately on Matrix:

This iteration doesn't implement history of kernel crashes, as that might be too complex to figure out. Instead it contains the Status feature will be able to detect if kdump configuration is up to date and if test of its configuration went succesfully. This feature was described at https://github.com/cockpit-project/cockpit/pull/19107#issuecomment-1680152357 .
 
![d6bf0579-a1ae-4780-b705-3325b7f9a11f](https://github.com/cockpit-project/cockpit/assets/42733240/f126a6a1-007a-4b66-a15c-1b4424e67c1e)
![8afcb9f7-8997-451f-990d-e84af08fb6d2](https://github.com/cockpit-project/cockpit/assets/42733240/f8a8f0ae-99db-42e7-96b3-1ebc2de57518)
![95765854-78c4-46ad-8470-3ab10fcdeaa3](https://github.com/cockpit-project/cockpit/assets/42733240/a66c18fe-3ac2-4678-8828-57a62566c33f)
![f0539452-acfe-49d1-b219-86ccc7735065](https://github.com/cockpit-project/cockpit/assets/42733240/a6fbd79e-55b1-4c94-9b69-d62e0a697c90)

**Comment by liutgnu** [comment]
Hi @garrett @skobyda,

Sorry for the late response. I'd like to share the working status about kdumpctl test/test-check.

I have planed to provide a simple interface from kexec-tools side as kdumpctl test/test-check as I mentioned previously. However after discussion, there is a trend for kexec-tools to be decoupled into sub-packages[1], so people don't like the idea to integrate more code into kexec-tools. Since kdumpctl test/test-check is not a functional feature, but a user experience feature. So I re-designed the test/test-check code by creating a new independent repo [2], and the interface will be changed into "kdump-test test/test-check".

"kdump-test test" will trigger a real kernel crash and dump, and "kdump-test test-check" will actually check if there is vmcore generated specifically for this test, it will automatically mount/ssh if the dump target is a remote fs based on kdump.conf, and give test result as "test successful" or "test failed". Only the "test successful" kdump configs can be regarded as valid.

If cockpit will continue to use the test/test-check interface, there will be an issue of cockpit how to integrate the kdump-test source code. Because we don't plan to make the new repo a rpm package to maintain for it is too small and simple. Cockpit can call test/test-check by invoke command line as "curl https://raw.githubusercontent.com/rhkdump/kdump-test/master/kdump-test{-lib.sh,} | bash -s test/test-check", instead of the previous "kdumpctl test/test-check". But it requires external network access, so it may be unfriendly for cases which can only access to internal netwok.

Anyway, this is not a final design and I'd like to hear your suggestions and comments if you are still interested in test/test-check interface.

Thanks,
Tao Liu

[1]: https://lists.fedoraproject.org/archives/list/kexec@lists.fedoraproject.org/message/5PFJMGBX42RQUDGZMCYMIMKN4RIFQMRC/
[2]: https://github.com/rhkdump/kdump-test
